pkg/chunk/cache/fifo_cache.go

```diff
@@ -179,91 +200,118 @@ func (c *FifoCache) Put(ctx context.Context, keys []string, values []interface{}
 	}
 }
 
+func (c *FifoCache) addToHead(entry *cacheEntry) {
+	if entry == nil {
+		return
+	}
+	entry.prev, entry.next = nil, c.first
+	if c.first != nil {
+		c.first.prev = entry
+	}
+	c.first = entry
+	if c.last == nil {
+		c.last = entry
+	}
+}
+
+func (c *FifoCache) deleteFromTail() *cacheEntry {
+	if c.last == nil {
+		return nil
+	}
+	ret := c.last
+	c.last = ret.prev
+	if c.last == nil {
+		c.first = nil
+	} else {
+		c.last.next = nil
+	}
+	return ret
+}
+
+func (c *FifoCache) deleteFromList(entry *cacheEntry) {
+	if entry == nil {
+		return
+	}
+	if entry.prev == nil {
+		c.first = entry.next
+	} else {
+		entry.prev.next = entry.next
+	}
+	if entry.next == nil {
+		c.last = entry.prev
+	} else {
+		entry.next.prev = entry.prev
+	}
+}
+
 func (c *FifoCache) put(ctx context.Context, key string, value interface{}) {
 	// See if we already have the entry
-	index, ok := c.index[key]
+	entry, ok := c.entries[key]
 	if ok {
-		entry := c.entries[index]
-		deltaSize := sizeOf(value) - sizeOf(entry.value)
-
-		entry.updated = time.Now()
-		entry.value = value
-
 		// Remove this entry from the FIFO linked-list.
-		c.entries[entry.prev].next = entry.next
-		c.entries[entry.next].prev = entry.prev
-
-		// Corner case: updating last element
-		if c.last == index {
-			c.last = entry.prev
-		}
+		c.deleteFromList(entry)
+		c.currSizeBytes -= sizeOf(entry)
+		delete(c.entries, key)
+		c.entriesCurrent.Dec()
+	}
 
-		// Insert it at the beginning
-		entry.next = c.first
-		entry.prev = c.last
-		c.entries[entry.next].prev = index
-		c.entries[entry.prev].next = index
-		c.first = index
+	entry = &cacheEntry{
+		updated: time.Now(),
+		key:     key,
+		value:   value,
+	}
+	entrySz := sizeOf(entry)
 
-		c.entries[index] = entry
-		c.memoryBytes.Add(float64(deltaSize))
+	if c.maxSizeBytes > 0 && entrySz > c.maxSizeBytes {
+		// Cannot keep this item in the cache
+		if ok {
+			// We do not replace the item
+			c.entriesEvicted.Inc()
+		}
+		c.memoryBytes.Set(float64(c.currSizeBytes))
 		return
 	}
-	c.entriesAddedNew.Inc()
+
+	if !ok {
+		c.entriesAddedNew.Inc()
+	}
 
 	// Otherwise, see if we need to evict an entry.
-	if len(c.entries) >= c.size {
+	for (c.maxSizeBytes > 0 && c.currSizeBytes+entrySz > c.maxSizeBytes) || (c.maxSizeItems > 0 && len(c.entries) >= c.maxSizeItems) {
 		c.entriesEvicted.Inc()
-		index = c.last
-		entry := c.entries[index]
-		deltaSize := sizeOf(key) + sizeOf(value) - sizeOf(entry.key) - sizeOf(entry.value)
-
-		c.last = entry.prev
-		c.first = index
-		delete(c.index, entry.key)
-		c.index[key] = index
-
-		entry.updated = time.Now()
-		entry.value = value
-		entry.key = key
-		c.entries[index] = entry
-		c.memoryBytes.Add(float64(deltaSize))
-		return
+		if evicted := c.deleteFromTail(); evicted != nil {
+			delete(c.entries, evicted.key)
+			c.currSizeBytes -= sizeOf(evicted)
+			c.entriesCurrent.Dec()
+		} else {
+			c.entries = nil
```

Comment:
In case we keep this `else` break for a good reason, then if you set it to `nil` it will panic
```suggestion
			c.entries = make(map[string]*cacheEntry)
```